src/java/org/apache/cassandra/index/sai/disk/v2/V2VectorIndexSearcher.java

```diff
@@ -297,37 +297,51 @@ private long getMaxSSTableRowId(PrimaryKeyMap primaryKeyMap, PartitionPosition r
-    {
-        return (int) (getObservedNodesRatio() * getRawExpectedNodes(limit, nPermittedOrdinals));
-    }
+        public CostEstimate(int nFilteredRows, int rawExpectedNodesVisited, int expectedNodesVisited)
+        {
+            assert rawExpectedNodesVisited >= 0 : rawExpectedNodesVisited;
+            assert expectedNodesVisited >= 0 : expectedNodesVisited;
 
-    /** the ratio of nodes visited by a graph search, to our estimate */
-    private double getObservedNodesRatio()
-    {
-        return actualExpectedRatio.getUpdateCount() >= 10 ? actualExpectedRatio.get() : 1.0;
+            this.nFilteredRows = nFilteredRows;
+            this.rawExpectedNodesVisited = rawExpectedNodesVisited;
+            this.expectedNodesVisited = expectedNodesVisited;
+        }
+
+        public boolean shouldUseBruteForce()
+        {
+            // ANN index will do a bunch of extra work besides the full comparisons (performing PQ similarity for each edge);
+            // brute force from sstable will also do a bunch of extra work (going through trie index to look up row).
+            // VSTODO I'm not sure which one is more expensive (and it depends on things like sstable chunk cache hit ratio)
+            // so I'm leaving it as a 1:1 ratio for now.
+            return nFilteredRows <= min(globalBruteForceRows, expectedNodesVisited);
+        }
+
+        public void updateStatistics(int actualNodesVisited)
+        {
+            assert actualNodesVisited >= 0 : actualNodesVisited;
+            actualExpectedNodesVisited.update(actualNodesVisited, rawExpectedNodesVisited);
+
+            if (actualNodesVisited >= 1000 && (actualNodesVisited > 2 * expectedNodesVisited || actualNodesVisited < 0.5 * expectedNodesVisited))
+                logger.warn("Predicted visiting {} nodes, but actually visited {} (observed:predicted ratio is {})",
+                            expectedNodesVisited, actualNodesVisited, (double) actualNodesVisited / expectedNodesVisited);
+        }
     }
 
-    private void updateExpectedNodes(int actualNodesVisited, int rawExpectedNodes)
+    private CostEstimate estimateCost(int limit, int nFilteredRows)
     {
-        assert rawExpectedNodes >= 0 : rawExpectedNodes;
-        assert actualNodesVisited >= 0 : actualNodesVisited;
-        double ratio = getObservedNodesRatio();
-        var expectedNodes = (int) (ratio * rawExpectedNodes);
-        if (actualNodesVisited >= 1000 && (actualNodesVisited > 2 * expectedNodes || actualNodesVisited < 0.5 * expectedNodes))
-            logger.warn("Predicted visiting {} nodes, but actually visited {} (observed:predicted ratio is {})",
-                        expectedNodes, actualNodesVisited, ratio);
-        actualExpectedRatio.updateAndGet(actualNodesVisited, rawExpectedNodes);
+        // update the raw expected value with a linear interpolation based on observed data
+        var interceptSlope = LinearFit.interceptSlopeFor(actualExpectedNodesVisited.getSnapshot().values);
```

Comment:
getSnopshot() copies the values array, can we avoid this?

Reply:
I don't think so, because unlike a normal Reservoir we care about getting matched pairs of values.  So we either need to copy or hold onto the synchronized lock for the entire operation